src/charm.py

```diff
@@ -121,92 +126,116 @@ def _get_database_relation_info(self) -> dict:
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring resources")
 
         self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
         logger.info("Pebble plan updated with new configuration, replanning")
-
         try:
             self._container.replan()
         except ChangeError as err:
             logger.error(str(err))
             self.unit.status = BlockedStatus("Failed to replan")
             return
+        finally:
+            if self.database.is_database_created():
+                if not self._container.exists(self._hydra_config_path):
+                    logger.info(
+                        "Config file not found. Pushing it to the container and replanning"
+                    )
+                    self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+                    self._container.replan()
+                self.unit.status = ActiveStatus()
+
+    def _on_database_created(self, event) -> None:
+        """Event Handler for database created event."""
+        logger.info("Retrieved database details")
 
-        try:
-            current_config = self._container.pull(self._hydra_config_path).read()
-        except (ProtocolError, PathError) as err:
-            logger.error(str(err))
-            self.unit.status = BlockedStatus(str(err))
-        else:
-            if current_config != self._config:
-                self._container.push(self._hydra_config_path, self._config, make_dirs=True)
-                logger.info("Updated hydra config, restarting the container")
-                self._container.restart()
-
-    def _update_container(self, event) -> None:
-        """Update configs, pebble layer and run database migration."""
-        self.unit.status = MaintenanceStatus("Configuring hydra container")
+        if not self.unit.is_leader():
+            # TODO: Observe leader_elected event
+            logger.info("Unit does not have leadership")
+            self.unit.status = WaitingStatus("Unit waiting for leadership to run the migration")
+            return
 
-        if not self.model.relations["pg-database"]:
-            # TODO: Observe relation_joined event instead of event deferring
+        if not self._container.can_connect():
             event.defer()
-            logger.error("Missing required relation with postgresql")
-            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+            logger.info("Cannot connect to Hydra container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        if not self.database.is_database_created():
+        self.unit.status = MaintenanceStatus(
+            "Configuring container and resources for database connection"
+        )
+
+        try:
+            self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
```

Comment:
Why so many items in the try/except. It's not obvious which calls can lead to which error.